sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/firestore/BaseFirestoreV1WriteFnTest.java

```diff
@@ -0,0 +1,640 @@
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.spy;
+
+import com.google.api.gax.grpc.GrpcStatusCode;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.ApiExceptionFactory;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.firestore.v1.BatchWriteRequest;
+import com.google.firestore.v1.BatchWriteResponse;
+import com.google.firestore.v1.Write;
+import com.google.rpc.Code;
+import com.google.rpc.Status;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.beam.sdk.io.gcp.firestore.FirestoreV1Fn.HasRpcAttemptContext;
+import org.apache.beam.sdk.io.gcp.firestore.FirestoreV1WriteFn.BaseBatchWriteFn;
+import org.apache.beam.sdk.io.gcp.firestore.FirestoreV1WriteFn.WriteElement;
+import org.apache.beam.sdk.io.gcp.firestore.RpcQos.RpcWriteAttempt.Element;
+import org.apache.beam.sdk.io.gcp.firestore.RpcQos.RpcWriteAttempt.FlushBuffer;
+import org.apache.beam.sdk.io.gcp.firestore.RpcQosImpl.FlushBufferImpl;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.joda.time.Instant;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+
+@SuppressWarnings("initialization.fields.uninitialized") // mockito fields are initialized via the Mockito Runner
+abstract class BaseFirestoreV1WriteFnTest<Out, Fn extends BaseBatchWriteFn<Out> & HasRpcAttemptContext> extends BaseFirestoreV1FnTest<Write, Out, Fn> {
+
+  protected static final Status STATUS_OK = Status.newBuilder().setCode(Code.OK.getNumber()).build();
+  @Mock(lenient = true)
+  protected BoundedWindow window;
+  @Mock
+  protected DoFn<Write, Out>.FinishBundleContext finishBundleContext;
+  @Mock
+  protected UnaryCallable<BatchWriteRequest, BatchWriteResponse> callable;
+  @Mock
+  protected RpcQos.RpcWriteAttempt attempt;
+  @Mock
+  protected RpcQos.RpcWriteAttempt attempt2;
+
+  @Before
+  public final void setUp() {
+    when(rpcQos.newWriteAttempt(any())).thenReturn(attempt, attempt2);
+
+    when(ff.getRpcQos(any()))
+        .thenReturn(rpcQos);
+    when(ff.getFirestoreRpc(pipelineOptions))
+        .thenReturn(rpc);
+    when(rpc.batchWriteCallable())
+        .thenReturn(callable);
+  }
+
+  @Override
+  @Test
+  public final void attemptsExhaustedForRetryableError() throws Exception {
+    Instant attemptStart = Instant.ofEpochMilli(0);
+    Instant rpc1Sta = Instant.ofEpochMilli(1);
```

Comment:
I think we can add two more letters to make it `rpc1Start`. :)

Reply:
Will update